Add reactive multipart request support

sdeleuze opened the pull request:
This draft PR is a first Multipart support implementation here for high-level feedbacks and discussion.

It does not include RequestMapping level support, integration tests do not work yet with Reactor and RxNetty and `NioMultipartResolver` need to be rewritten using `Flux.create()` instead of `ReplayProcessor`.

sdeleuze commented:
I have updated this PR witgh the following changes:
 - Tests now run fine with Reactor Netty and RxNetty
 - `getAllParts()` now caches the resulting mono to allow multiple invocation
 - There is a single `Part` object suitable usable for file and non file parts
 - We now use `Flux.create()` instead of `ReplayProcessor`
 - `NioPart#transferTo()` now uses zero copy  when possible (when a temporary file has been previously created)

@rstoyanchev Could you please make a high-level review of this commit?

rstoyanchev commented:
@sdeleuze, apologies for suggesting it but ServerHttpRequest doesn't seem like the right level for exposing multipart support. Having MultipartResolver take ServerHttpRequest as input is already an indication that it's a higher level concern. Then MultipartResolver has to be injected into every request implementation, which is rather repetitive and artificial. It's also a cyclical dependency even at the package level since the http package is logically at a level below the web package and should not depend on anything from it.

It seems to be a better fit for being exposed at the ServerWebExchange level and to avoid a circular dependency again, MultipartResolver and Part should be in server while with maybe a server.multipart sub-package for the implementations. For initialization a MultipartResolver could be a field of HttpWebHandlerAdapter and passed into every created ServerWebExchange. This is very similar to WebSessionManager and WebSession.

For the caching, the cache operator should really be called at construction time or otherwise concurrent calls to getParts can each create a separate pipeline. See how DefaultWebExchange does it for WebSession.

For when the request is not multipart, could we be more lenient and return an empty stream from getParts and a Mono with empty map from getAllParts? It removes the need to check isMultipartRequest first.

Do we need Part.getContentType if we have getHeaders since there are plenty of methods already on part? 

I haven't looked into the implementation library but I presume it is non-blocking indeed? It seems to be returning java.io.InputStream.

sdeleuze commented:
No worry indeed, I agree that `ServerHttpRequest` is not the right level to implement that for the reasons you mentioned. I also agree with your remarks on caching and being more lenient if the request is not multipart.

We can remove `Part.getContentType()` since this is just a shortcut to `Part.getHeaders().getContentType()` (with maybe a Javadoc hint to help users for file parts).

Currently I think JSON parts will be natively treated by the underlying library like a part with a `String` based value since it has no filename header. Even if in `NioPart` I allowed to get the content both as a `String` or as a `Publisher<DataBuffer>`, that could be not optimal.

Not sure if that makes sens, but should we add some `HttpMessageReader` based capabilities with some `<T> Flux<T> getContentAsStream(Class<T> type)` and  `<T> Mono<T> getContentAs(Class<T> type)` methods available on `Part` ? Or should we delegate that part to a composite `MultipartHttpMessageReader` that could be initialized with a `List<HttpMessageReader>` like we do on the writing side for SSE?

About the NIO Multipart library, it seems to me that the request parsing is done in a non blocking way but sadly we can only access to data as `String` or `InputStream` depending on the part type, so it may be blocking I suppose. I have implemented an optimisation to be able to perform a zero-copy transfer if the `InputStream` is a `FileInputStream`.

I can take care of the changes discussed here if you are ok.

I am going to create an issue on their GitHub to ask them more insight about the blocking behavior implied by `InputStream` and that fact that non-file part content is only accessible as a String.

rstoyanchev commented:
Okay, I'm not too sure what you mean with the HttpMessageReader based capabilities. I would think that Part should only expose basic ways of getting to the data, but once you've updated the PR maybe it will be more clear.

sdeleuze commented:
I have pushed an updated PR, and asked [a new question](https://github.com/synchronoss/nio-multipart/issues/4#issuecomment-254652854) to NIO Multipart project lead about how to perform non-blocking processing with their current `InputStream` / `OutputStream` based API.

sdeleuze commented:
@rstoyanchev Could you please review the last version of this PR?

rstoyanchev commented on spring-web/src/main/java/org/springframework/web/multipart/reactive/nio/NioMultipartResolver.java:
```diff
@@ -0,0 +1,167 @@
+				}
+				catch (IOException ex) {
+					throw Exceptions.bubble(e);
+				}
```
Should this notify the parser about the error which calls `error` on the emitter, so that subscribers for the parts will get the error?

rstoyanchev commented on spring-web/src/main/java/org/springframework/web/multipart/reactive/nio/NioMultipartResolver.java:
```diff
@@ -0,0 +1,167 @@
+				public void onError(String message, Throwable cause) {
+					emitter.error(new MultipartException(message, cause));
+				}
+			});
```
An inner class perhaps given the number of methods?

rstoyanchev commented on spring-web/src/main/java/org/springframework/web/server/ServerWebExchange.java:
```diff
@@ -118,4 +122,19 @@
+	 * or a {@link Mono} emitting an error the current request is not a multipart one.
+	 */
+	Mono<MultiValueMap<String, Part>> getAllParts();
+
```
These could be ordered right below `getPrincipal()` above with `getAllParts` ahead of `getParts` as the more common case. Same for decorator and default implementations.

sdeleuze replied:
I changed the order and also proposed new names in order to be more consistent with `getFormData()`: `Mono<MultiValueMap<String, Part>> getFormParts()` and `Flux<Part> getFormPartsAsFlux()`. Are you ok with these new names?

rstoyanchev commented on spring-web/src/test/java/org/springframework/http/server/reactive/MultipartIntegrationTests.java:
```diff
@@ -0,0 +1,164 @@
+						else {
+							fail();
+						}
+					})
```
Could `exchange.getParts()` with `collectToMap` be used to get all parts first and then invoke a method that asserts all parts, which can also be called from `assertGetAllParts`?

rstoyanchev reviewed (changes requested):
Having `ServerWebExchange` to expose multipart methods and delegate to `MultipartResolver` internally looks alright. In fact the MultipartResolver seems to belong in the http package. It's like an HttpMessageReader and could even be used with an HttpMessageReader to further parse the content of a part. I suggest moving `web.multipart.reactive` to `http.reactive` right next to the codec package.

sdeleuze commented on spring-web/src/main/java/org/springframework/web/multipart/reactive/nio/NioMultipartResolver.java:
```diff
@@ -0,0 +1,167 @@
+					MultiValueMap<String, Part> partsMap = new LinkedMultiValueMap<>();
+					parts.forEach(part -> partsMap.add(part.getName(), part));
+					return partsMap;
+				});
```
Reactor `collectMultimap()` is using a `Collection` for the value type while `LinkedMultiValueMap` is taking a `Map<K, List<V>>` parameter, that's why I did not use it in my first implementation, but after more thoughts I was able to use it in my updated commit thanks to `Collectors.toMap()` to change the value type of the map.

sdeleuze commented:
@rstoyanchev  I took in account your feedbacks and:
 - Changed the package to `org.springframework.http.codec.multipart`
 - Polished the Javadoc
 - Renamed `getAllParts()` to `getFormParts()`
 - Renamed `getParts()` to `getFormPartsAsFlux()`
 - Integrated the `MultipartHttpMessageWriter` commit

sdeleuze committed:
Add reading reactive multipart request support

This commit introduces reactive multipart support by adding a new
MultipartHttpMessageReader interface (with default methods) and a
SynchronossMultipartHttpMessageReader implementation based on
the Synchronoss NIO Multipart implementation
(https://github.com/synchronoss/nio-multipart).

Issue: SPR-14546

sdeleuze committed:
Add MultipartHttpMessageWriter

This commit adds a reactive HttpMessageWriter that allows
to write multipart HTML forms with multipart/form-data
media type.

Issue: SPR-14546

sdeleuze committed:
Add multipart support to WebFlux functional API

Issue: SPR-14546

sdeleuze committed:
Expose ServerCodecConfigurer as a bean

With this commit, ServerCodecConfigurer is now exposed as a bean in
order to be provided to DefaultServerWebExchange via
WebHttpHandlerBuilder and HttpWebHandlerAdapter. This allows
DefaultServerWebExchange to get configured codecs for reading form or
multipart requests.

Issue: SPR-14546

sdeleuze committed:
Add multipart support to ServerWebExchange

Issue: SPR-14546

sdeleuze committed:
Add @Request(Param/Part) support for multipart requests

Issue: SPR-14546

rstoyanchev commented:
This is now in master.
